### PR TITLE
[#12920] Create script to migrate noSQL test data to SQL schema format

### LINKED ISSUE #12920: Existing Test data json has to be manually migrated to SQL format
**Description of feature/enhancement**
Existing JSON test data used for tests is in the format for the current noSQL database. Migrating each JSON file by hand can be tedious due to the foreign key relationships enforced by SQL and any other attribute changes. This can be solved by using a user script to migrate these files instead

**Proposed Usage**
1. Specify JSON file to migrate, and create migrated file in same directory as JSON file
2. Create a folder to place JSON files to migrate and migrate all JSON files

Comments on what would be easier to use to use is welcome

### PR DESCRIPTION
**Pull request by NicolasCwy**
Fixes #12920 

**Outline of Solution**
Create user script which can be run to convert noSQL test data (e.g E2E test data, typicalDatabundle.json) into the SQL format to respect foreign key references using the GSON library which converts java objects to JSON. I have tested this on a few migrated E2E tests without any issue.

**Note**
Since GSON was used to convert the Java objects to JSON, the attributes are sorted by the order which the Java SQL entities are, which might make comparison of the converted JSON to the current test data difficult (see below)

**Usage**
`./gradlew execScript -PuserScript="testdataconversion/ConvertDatastoreJsonToSqlJson" --args="JSON_FILE_PATH_HERE"`

The converted file is output in the same directory as the JSON file, and named  `YourFileNameSql.json`. Comments on how to make this more user friendly are welcome

**Images**
JSON Output
image

Instructory SQL Entity Definition
image

### CODE REVIEW

File: src/client/java/teammates/client/scripts/testdataconversion/toposort.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
+#Python program to print topological sorting of a DAG
```
Comment (cedricongjh) [inline src/client/java/teammates/client/scripts/testdataconversion/toposort.py:1]: wondering if this file is needed? not sure if i see it being used anywhere in the .java script

File: src/client/java/teammates/client/scripts/testdataconversion/ConvertDatastoreJsonToSqlJson.java [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+import teammates.storage.sqlentity.Instructor;
+import teammates.storage.sqlentity.Notification;
+import teammates.storage.sqlentity.ReadNotification;
+import teammates.storage.sqlentity.Section;
+import teammates.storage.sqlentity.Student;
+import teammates.storage.sqlentity.Team;
+import teammates.test.FileHelper;
+
+/**
+ * Class to create JSON test data in SQL format from a noSQL JSON file.
```
Comment (cedricongjh) [inline src/client/java/teammates/client/scripts/testdataconversion/ConvertDatastoreJsonToSqlJson.java:30]: should we add a guide to the usage here in this comment?

File: src/client/java/teammates/client/scripts/testdataconversion/ConvertDatastoreJsonToSqlJson.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+            FeedbackResponseComment sqlFeedbackResponseComment = entityConverter.convert(feedbackReponseComment);
+            sqlDataBundle.feedbackResponseComments.put(k, sqlFeedbackResponseComment);
+        });
+    }
+
+    public static void main(String[] args) throws IOException, InvalidParametersException {
+        if (args.length > 0) {
+            File inputFile = new File(args[0]);
+            ConvertDatastoreJsonToSqlJson script = new ConvertDatastoreJsonToSqlJson(inputFile);
+            String outputFileName = "output.json";
```
Comment (cedricongjh) [inline src/client/java/teammates/client/scripts/testdataconversion/ConvertDatastoreJsonToSqlJson.java:208]: could we have the outputFileName be <name of .json>SQLTest.json, (e.g. For InstructorHomePageE2ETest.json, have it output: InstructorHomePageE2ESqlTest.json or InstructorHomePageE2ETestSql.json (this is easier to do))

**Review by cedricongjh — CHANGES_REQUESTED** [review]
thanks for working on this!

found a bug with teams: 
image

for `InstructorHomePageE2ETest.json`, the teams appear to be duplicated (in NoSQL, same team name under the same course means the same team)

File: src/client/java/teammates/client/scripts/testdataconversion/toposort.py (hunk 1 above)
Comment (NicolasCwy) [reply src/client/java/teammates/client/scripts/testdataconversion/toposort.py:1]: I was wondering if I should leave some artifact of how I decided the order on how to migrate dependent entities but if you think its not needed I can remove this. Perhaps I'll just add it as a small comment in the actual code

**Comment by NicolasCwy** [comment]
@cedricongjh It was a bug! I used the wrong JSON key for teams. Have fixed this

**Review by cedricongjh — APPROVED** [review]
LGTM, thanks for the work on this!

**Review by weiquu — APPROVED** [review]
LGTM

**Pull request merged**